artifact.go

```diff
@@ -203,67 +213,423 @@ func (m artifactModule) getBuildTestsResultsHandler(c *gin.Context) {
+	if !ok {
+		return
+	}
+
+	summaries := []TestResultSummary{}
+	err := m.Database.
+		Preload("Artifact").
+		Where(&TestResultSummary{BuildID: uint(buildID)}).
+		Find(&summaries).
+		Error
+
+	if err != nil {
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result summaries for build with ID %d from database.",
+			buildID))
+		return
+	}
+
+	summaryList := TestResultListSummary{
+		BuildID:   buildID,
+		Summaries: summaries}
+
+	for _, v := range summaries {
+		summaryList.Failed += v.Failed
+		summaryList.Passed += v.Passed
+		summaryList.Skipped += v.Skipped
+	}
+
+	summaryList.Total = summaryList.Failed + summaryList.Passed + summaryList.Skipped
+
+	c.JSON(http.StatusOK, summaryList)
+}
+
+func parseMultipartFormData(c *gin.Context, buildID uint) ([]file, bool) {
+	files := make([]file, 0)
+
+	form, err := c.MultipartForm()
+	if err != nil {
+		ginutil.WriteMultipartFormReadError(c, err,
+			fmt.Sprintf("Failed reading multipart-form content from request body when uploading new"+
+				" artifact for build with ID %d.", buildID))
+		return nil, false
+	}
+
 	for k := range form.File {
 		if fhs := form.File[k]; len(fhs) > 0 {
-			f, err := fhs[0].Open()
-			if err != nil {
-				ginutil.WriteMultipartFormReadError(c, err, fmt.Sprintf(
-					"Failed with starting to read file content from multipart form request body when uploading new artifact for build with ID %d.",
-					buildID))
-				return
+			data, ok := readMultipartFileData(c, buildID, fhs[0])
+			if !ok {
+				return files, false
 			}
 
-			data, err := ioutil.ReadAll(f)
-			if err != nil {
-				ginutil.WriteMultipartFormReadError(c, err, fmt.Sprintf(
-					"Failed reading file content from multipart form request body when uploading new artifact for build with ID %d.",
-					buildID))
-				return
-			}
+			files = append(files, file{
+				name:     k,
+				fileName: fhs[0].Filename,
+				data:     data,
+			})
+		}
+	}
 
-			artifact := Artifact{
-				Data:     data,
-				Name:     k,
-				FileName: fhs[0].Filename,
-				BuildID:  buildID,
-			}
+	return files, true
+}
 
-			err = m.Database.Create(&artifact).Error
-			if err != nil {
-				ginutil.WriteDBWriteError(c, err, fmt.Sprintf(
-					"Failed saving artifact with name %q for build with ID %d in database.",
-					artifact.FileName, buildID))
-				return
-			}
+func readMultipartFileData(c *gin.Context, buildID uint, fh *multipart.FileHeader) ([]byte, bool) {
+	if fh == nil {
+		return nil, false
+	}
 
-			log.Info().
-				WithString("filename", artifact.Name).
-				WithUint("build", buildID).
-				WithUint("artifact", artifact.ArtifactID).
-				Message("File saved as artifact")
+	f, err := fh.Open()
+	if err != nil {
+		ginutil.WriteMultipartFormReadError(c, err,
+			fmt.Sprintf("Failed with starting to read file content from multipart form request body when"+
+				" uploading new artifact for build with ID %d.", buildID))
+		return nil, false
+	}
+	defer f.Close()
```

Comment:
Can actually also fail during closing. Don't know the best way to handle that, but I think we should at least log it

```suggestion
	defer func() {
		if err := f.Close(); err != nil {
			log.Error().
				WithError(err).
				Message("Failed to close multipart form request body file handle.")
		}
	}
```

Reply:
Looking closer at the source, the implementation of `Close()` for the file returned is just returning nil, so we can remove it completely. Although this is only true when the `multipart.FileHeader`'s content property is non-nil. I made sure to check that that was the case here.

Reply:
While this is true now, it could change later on. To follow "best coding practices", you should always at least close a file handle when you're done with it, even if the underlying implementation does nothing in this version.

Suggest at least reverting commit 64181eb